storage/innobase/read/read0read.cc

```diff
@@ -262,4 +307,552 @@ void trx_sys_t::clone_oldest_view(ReadViewBase *view) const
+}
+
+void SCN_Mgr::batch_write(buf_block_t *block, mtr_t *mtr)
+{
+  ut_ad(srv_use_scn);
+  LazyCleanoutRecs lrecs;
+  lrecs.clear();
+  block->copy_and_free(lrecs);
+
+  for (auto &lrec : lrecs)
+  {
+    rec_t *rec= lrec.first;
+    ulint trx_id_offset= std::get<0>(lrec.second);
+    trx_id_t trx_id= std::get<1>(lrec.second);
+    trx_id_t scn= std::get<2>(lrec.second);
+    set_scn(mtr, block, rec, trx_id_offset, trx_id, scn);
+  }
+}
+
+void SCN_Mgr::cleanout_task(uint32_t slot)
+{
+  ut_a(srv_use_scn);
+  CleanoutWorker *worker= m_cleanout_workers[slot];
+  MDL_ticket *mdl_ticket;
+  THD *thd= worker->get_thd();
+  while (!m_abort.load())
+  {
+    PageSets pages;
+    pages.clear();
+    worker->take_pages(pages);
+    if (pages.empty())
+    {
+      break;
+    }
+    /* Process pages that need to be modified */
+    for (auto &val : pages)
+    {
+      if (m_abort.load())
+      {
+        break;
+      }
+      page_id_t page_id(val.first);
+      table_id_t table_id= val.second;
+      mdl_ticket= nullptr;
+      /* Prevents the tablespace from being dropped during the operation */
+      dict_table_t *table= dict_table_open_on_id(
+          table_id, false, DICT_TABLE_OP_NORMAL, thd, &mdl_ticket);
+      if (table == nullptr)
+      {
+        continue;
+      }
+      fil_space_t *space= fil_space_get(page_id.space());
+      ulint zip_size= 0;
+      ut_a(space);
+      zip_size= space->zip_size();
+      mtr_t mtr;
+      buf_block_t *block= nullptr;
+      mtr_start(&mtr);
+      mtr.set_in_scn_cleanout();
+      block= buf_page_get_gen(page_id, zip_size, RW_X_LATCH, nullptr,
+                              BUF_GET_IF_IN_POOL, &mtr, nullptr);
+      if (block != nullptr)
+      {
+        if (!table->is_active_ddl())
+        {
+          mtr.set_named_space(space);
+          batch_write(block, &mtr);
+        }
+        else
+        {
+          block->clear_cursor();
+        }
+      }
+      mtr_commit(&mtr);
+      dict_table_close(table, false, thd, mdl_ticket);
+    }
+  }
+}
+
+void run_cleanout_task(void *arg)
+{
+  uint32_t slot= *(uint32_t *) arg;
+  scn_mgr->cleanout_task(slot);
+}
+
+void SCN_Mgr::cleanout_task_monitor()
+{
+  for (uint32_t i= 0; i < m_max_cleanout_threads; i++)
+  {
+    auto cleanout_work= m_cleanout_workers[i];
+    if (!cleanout_work->is_empty() && !cleanout_work->is_running())
+    {
+      srv_thread_pool->submit_task(cleanout_work->get_task());
+    }
+  }
+}
+
+void cleanout_task_monitor(void *) { scn_mgr->cleanout_task_monitor(); }
+
+void run_view_task(void *) { scn_mgr->view_task(); }
+
+void SCN_Mgr::start()
+{
+  if (!srv_use_scn || !m_cleanout_workers)
+  {
+    return;
+  }
+  m_abort= false;
+
+  m_cleanout_task_timer.reset(
+      srv_thread_pool->create_timer(::cleanout_task_monitor, nullptr));
+  m_cleanout_task_timer->set_time(0, 1000);
+
+  m_view_task_timer.reset(
+      srv_thread_pool->create_timer(run_view_task, nullptr));
+  m_view_task_timer->set_time(0, 1000);
```

Comment:
What are these for? The commit message does not mention any periodic task or timer.

Could any of this be executed by the `purge_coordinator_task`?

Reply:
To optimize the performance of MVCC, we need to write back SCN to TRX_ID fields of rows that touched by reading. These write back operations are done in background tasks. The timer `m_view_task_timer` is used to traverse `trx_sys_t::rw_trx_hash ` periodically and refresh `SCN_Mgr::m_safe_limit_no` and `SCN_Mgr::m_min_active_id`.
The work of these two timers can be done in `purge_coordinator_task`, but it may make the code become complex.